// ngx-file-drop/esm5/ngx-file-drop/ngx-file-drop.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxFileDropComponent } from './ngx-file-drop.component';
import { NgxFileDropContentTemplateDirective } from './ngx-templates.directive';
var NgxFileDropModule = /** @class */ (function () {
    function NgxFileDropModule() {
    }
    NgxFileDropModule = __decorate([
        NgModule({
            declarations: [
                NgxFileDropComponent,
                NgxFileDropContentTemplateDirective,
            ],
            imports: [
                CommonModule
            ],
            exports: [
                NgxFileDropComponent,
                NgxFileDropContentTemplateDirective,
            ],
            providers: [],
            bootstrap: [
                NgxFileDropComponent
            ],
        })
    ], NgxFileDropModule);
    return NgxFileDropModule;
}());
export { NgxFileDropModule };
